### PR TITLE
Bug 1131727 - Create testdroid client

### PR DESCRIPTION
**Pull request by gregarndt**

### CODE REVIEW

File: src/examples/adb_marionette_proxy.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  let baseUrl = process.argv[2];
+  let username = process.argv[3];
+  let password = process.argv[4];
+
+  if (process.argv.length < 5) {
+    console.log("Must supply url, username, and password");
+    process.exit(-1);
+  }
+  let session;
+  let t;
```
Comment (lightsofapollo) [inline src/examples/adb_marionette_proxy.js]: nit: use more descriptive name

File: src/examples/adb_marionette_proxy.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import Testdroid from 'testdroid-client';
+let session;
+async () => {
+  let baseUrl = process.argv[2];
+  let username = process.argv[3];
+  let password = process.argv[4];
+
+  if (process.argv.length < 5) {
+    console.log("Must supply url, username, and password");
+    process.exit(-1);
```
Comment (lightsofapollo) [inline src/examples/adb_marionette_proxy.js]: nit: why -1 instead of 1?

File: src/examples/adb_marionette_proxy.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import Testdroid from 'testdroid-client';
+let session;
+async () => {
+  let baseUrl = process.argv[2];
```
Comment (lightsofapollo) [inline src/examples/adb_marionette_proxy.js]: Why don't you use argparse here?

File: src/examples/adb_marionette_proxy.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      let adb = await t.getProxy('adb', session.id);
+      let marionette = await t.getProxy('marionette', session.id);
+      console.log(marionette);
+    }
+    else {
+      console.log("Could not find online device");
+    }
+  }
+  catch (e) {
+    console.log(e);
```
Comment (lightsofapollo) [inline src/examples/adb_marionette_proxy.js:30]: nit: can you explain the logic around this try/catch a bit it looks like this is designed to handle the case where you have a session but maybe some commands fail? Might not need this at all...

File: src/examples/device_in_label_group.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import Testdroid from 'testdroid-client';
+
+async () => {
+  let baseUrl = process.argv[2];
```
Comment (lightsofapollo) [inline src/examples/device_in_label_group.js]: nit: see above also might be worth seeing what I did in tc-vcs the pattern here looks very similar.

File: src/lib/project.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    debug(`Deleting param '${parameter.key}' for test run '${testRun.id}'`);
+    let res = await this.client.del(`${this.URI}/runs/${testRun.id}/config/parameters/${parameter.id}`);
+
+    if (!res.ok) {
+      let err = `Could not delete param '${parameter.key}'. ${res.error.message}`;
+      debug(err);
+      throw new Error(err);
+    }
+
+    return;
```
Comment (lightsofapollo) [inline src/lib/project.js]: nit unsed

File: src/lib/test_run.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  /**
+   * Aborts the test run
+   *
+   * @returns {Object} response object
+   */
+  async abort() {
+    let res = await this.client.post(`${this.URI}/abort`);
+    console.log(res);
```
Comment (lightsofapollo) [inline src/lib/test_run.js]: nit: unprefixed console log in lib looking code

File: src/lib/testdroid.js [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+      r.query(payload);
+    } else {
+      r.send(payload);
+    }
+
+    var res = await r.end();
+    return res;
+  }
+
+  async buildHeaders(additionalHeaders) {
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: nit use default arguments?

File: src/lib/testdroid.js [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    let endpoint = urljoin(this.apiUrl, path);
+    opts = typeof(opts) === 'undefined' ? {} : opts;
+    let payload = 'payload' in opts ? opts.payload : {};
+    let headers = await this.buildHeaders(opts.headers);
+    let r = request(method.toUpperCase(), endpoint);
+
+    r.set(headers);
+
+    if (method.toUpperCase() === 'GET') {
+      r.query(payload);
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: nit: (Can't help myself) use longer name?

File: src/lib/testdroid.js [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+   * __request('get', '/devices', requestOptions);
+   *
+   * @param {String} method - method for the request.  'get' or 'post'
+   * @param {String} path - endpoint to submit request to
+   * @param {opts} opts - optional opts to include.  Maybe include payload and/or headers
+   * @returns {Object} Response
+   */
+  async __request(method, path, opts) {
+    let endpoint = urljoin(this.apiUrl, path);
+    opts = typeof(opts) === 'undefined' ? {} : opts;
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: nit: default arguments?

File: src/lib/testdroid.js [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+   * };
+   *
+   * __request('get', '/devices', requestOptions);
+   *
+   * @param {String} method - method for the request.  'get' or 'post'
+   * @param {String} path - endpoint to submit request to
+   * @param {opts} opts - optional opts to include.  Maybe include payload and/or headers
+   * @returns {Object} Response
+   */
+  async __request(method, path, opts) {
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: This is a longer comment about the design here... I tend to put what I consider "private" (or more specifically this code is just helpers that have nothing to do with this object) as stand alone functions so it is clear that nobody else should call them and so it is clear they don't depend on the state of this class/object/thing

File: src/lib/testdroid.js [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+   * Examples:
+   * // Get request using payload as object
+   * client.get('/devices', { 'payload': { 'limit': 1 } });
+   *
+   * @param {String} path - API endpoint to submit post request to.
+   * @param {Object} opts - Payload to send.
+   * * @returns {Object} Response
+   */
+  async get(path, opts) {
+    opts = typeof(opts) !== 'undefined' ? opts : {};
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: nit: default opts

File: src/lib/testdroid.js [hunk 13]
```diff
@@ -0,0 +1,455 @@
...
+      r.send(payload);
+    }
+
+    var res = await r.end();
+    return res;
+  }
+
+  async buildHeaders(additionalHeaders) {
+    let headers = typeof(additionalHeaders) !== 'undefined' ? additionalHeaders : {};
+    let token = await this.getToken();
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: Ah I see... This is a minor thing but I would probably put the private api stuff in another object (or more likely a factory for request objects)

File: src/lib/testdroid.js [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+  /**
+   * Retrieves all devices matching a given name.
+   * @param {String} deviceName - Name of the device
+   * @return {Object} Device information
+   */
+  async getDevicesByName(deviceName) {
+    let devices = await this.getDevices();
+
+    var matchedDevices = [];
+    for (var i=0; i < devices.length; i++) {
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: nit: use filter instead of for loop

File: src/lib/testdroid.js [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+    };
+
+    let res;
+    let maxRetries = 30;
+    let queryFormat = "{\"type\":\"%s\", \"sessionId\": %d}";
+    // Attempt to get a proxied adb/marionette connection.  Stop after 150 seconds
+    for (let i = 1; i <= maxRetries; i++) {
+      debug(`Creating proxied '${type}' session. Attempt ${i} of ${maxRetries}`);
+      let encodedQuery = encodeURIComponent(util.format(queryFormat, type, sessionId));
+      res = await this.get(util.format('proxy-plugin/proxies?where=%s', encodedQuery));
```
Comment (lightsofapollo) [inline src/lib/testdroid.js]: heh json encoded into get params interesting...

File: src/examples/adb_marionette_proxy.js (hunk 4 above)
Comment (gregarndt) [inline src/examples/adb_marionette_proxy.js:30]: Yea, on line 19 a device session is started and if getting the proxy for adb and marionette fail, I wanted the dession to be cleaned up.

File: src/examples/adb_marionette_proxy.js (hunk 1 above)
Comment (gregarndt) [inline src/examples/adb_marionette_proxy.js]: fixed with c6ecfd2

File: src/lib/project.js (hunk 6 above)
Comment (gregarndt) [inline src/lib/project.js]: Fixed with c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/examples/adb_marionette_proxy.js (hunk 2 above)
Comment (gregarndt) [inline src/examples/adb_marionette_proxy.js]: fixed with c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/examples/adb_marionette_proxy.js (hunk 3 above)
Comment (gregarndt) [inline src/examples/adb_marionette_proxy.js]: Updated with 82f835c70e8378013db9998cfd7ff148e9e06881 Probably should make a higher level command that could be used for more things than just the examples in the future.

File: src/lib/test_run.js (hunk 7 above)
Comment (gregarndt) [inline src/lib/test_run.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/lib/testdroid.js (hunk 8 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/lib/testdroid.js (hunk 9 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/lib/testdroid.js (hunk 10 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/lib/testdroid.js (hunk 11 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in 28d82355c17ff40e65e0456934a4a1bedf380b71

File: src/lib/testdroid.js (hunk 12 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

File: src/lib/testdroid.js (hunk 13 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in 28d82355c17ff40e65e0456934a4a1bedf380b71

File: src/lib/testdroid.js (hunk 14 above)
Comment (gregarndt) [inline src/lib/testdroid.js]: fixed in c6ecfd2f8285608d9c1935b0f08273294108dbce

**Pull request closed without merging**